File: backend/app/models/game.py
"""
Database models for snooker scoring
FRD Section 13: Player, Game, Event tables
"""

from datetime import datetime
from uuid import uuid4

from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import relationship

from app.db import Base


def generate_uuid():
    return str(uuid4())


class Player(Base):
    __tablename__ = "players"

    id = Column(UUID(as_uuid=False), primary_key=True, default=generate_uuid)
    name = Column(String(30), nullable=False)
    game_id = Column(UUID(as_uuid=False), ForeignKey("games.id"), nullable=False)
    team_id = Column(UUID(as_uuid=False), ForeignKey("teams.id"), nullable=True)
    score = Column(Integer, default=0)

    game = relationship("Game", back_populates="players")
    team = relationship("Team", back_populates="players")
    events = relationship("GameEvent", back_populates="player")


class Team(Base):
    __tablename__ = "teams"

    id = Column(UUID(as_uuid=False), primary_key=True, default=generate_uuid)
    name = Column(String(50), nullable=False)
    game_id = Column(UUID(as_uuid=False), ForeignKey("games.id"), nullable=False)
    score = Column(Integer, default=0)

    game = relationship("Game", back_populates="teams")
    players = relationship("Player", back_populates="team")


class Game(Base):
    __tablename__ = "games"

    id = Column(UUID(as_uuid=False), primary_key=True, default=generate_uuid)
    mode = Column(String(20), default="singles")  # singles | doubles
    current_player_index = Column(Integer, default=0)
    reds_remaining = Column(Integer, default=15)
    current_break = Column(Integer, default=0)
    last_potted_color = Column(String(20), nullable=True)
    status = Column(String(20), default="active")  # setup | active | finished
    start_time = Column(DateTime, default=datetime.utcnow)

    players = relationship("Player", back_populates="game", cascade="all, delete-orphan")
    teams = relationship("Team", back_populates="game", cascade="all, delete-orphan")
    events = relationship("GameEvent", back_populates="game", cascade="all, delete-orphan")


class GameEvent(Base):
    __tablename__ = "events"

    id = Column(UUID(as_uuid=False), primary_key=True, default=generate_uuid)
    game_id = Column(UUID(as_uuid=False), ForeignKey("games.id"), nullable=False)
    player_id = Column(UUID(as_uuid=False), ForeignKey("players.id"), nullable=False)
    event_type = Column(String(20), nullable=False)  # pot | foul
    ball = Column(String(20), nullable=True)
    points = Column(Integer, nullable=False)
    description = Column(Text, nullable=True)
    timestamp = Column(DateTime, default=datetime.utcnow)

    game = relationship("Game", back_populates="events")
    player = relationship("Player", back_populates="events")

File: frontend/src/types/game.ts
/**
 * Shared types for the Smart Snooker Scoring System
 */

export type BallType = 'red' | 'yellow' | 'green' | 'brown' | 'blue' | 'pink' | 'black'

/** Ball involved in foul - includes cue ball (potted white) */
export type FoulBallType = BallType | 'cue_ball'

/** Ball on = ball that must be struck first (CR-001) */
export type BallOnType = 'red' | 'colour' | BallType

export const BALL_VALUES: Record<BallType, number> = {
  red: 1,
  yellow: 2,
  green: 3,
  brown: 4,
  blue: 5,
  pink: 6,
  black: 7,
}

/** Color sequence when reds are gone (FRD 3.3) */
export const FINAL_COLOR_SEQUENCE: BallType[] = [
  'yellow',
  'green',
  'brown',
  'blue',
  'pink',
  'black',
]

export interface Player {
  id: string
  name: string
  score: number
  teamId?: string
}

export interface Team {
  id: string
  name: string
  playerIds: string[]
  score: number
}

export type GameEventType = 'pot' | 'foul'

export interface GameEvent {
  id: string
  type: GameEventType
  playerId: string
  ball?: BallType
  points: number
  timestamp: string
  description?: string
  /** Foul metadata (CR-001) */
  ballOn?: BallOnType
  ballInvolved?: FoulBallType
}

export interface GameState {
  id: string
  mode: 'singles' | 'doubles'
  players: Player[]
  teams?: Team[]
  currentPlayerIndex: number
  redsRemaining: number
  currentBreak: number
  lastPottedColor?: BallType
  /** Ball that must be struck first (CR-001) */
  ballOn?: BallOnType
  events: GameEvent[]
  status: 'setup' | 'active' | 'finished'
  createdAt: string
}
